ts/client/services/rpc.ts

```diff
@@ -20,12 +22,70 @@ export class RPC {
         fastify: FastifyInstance,
         transferPool?: ITransferPool
     ) {
+        fastify.register(cors, {
+            origin: (origin, cb) => {
+                if (/localhost/.test(origin)) {
+                    //  Request from localhost will pass
+                    cb(null, true);
+                    return;
+                }
+                // Generate an error on other origins, disabling access
+                cb(new Error("Not allowed"), false);
+            }
+        });
         fastify.get<{ Params: { stateID: number } }>(
             "/user/state/:stateID",
-            async function(request) {
-                const stateID = request.params.stateID;
-                const state = await l2Storage.state.get(stateID);
-                return state.toJSON();
+            async function(request, reply) {
+                try {
+                    const { stateID } = request.params;
+                    const state = await l2Storage.state.get(stateID);
+                    return state.toJSON();
+                } catch (error) {
+                    return reply.status(404).send({ error: "state not found" });
+                }
+            }
+        );
+        fastify.get<{ Params: { pubkeyHash: string } }>(
+            "/user/state/pubkey/:pubkeyHash",
+            async function(request, reply) {
+                try {
+                    const { pubkeyHash } = request.params;
+                    const stateIndices = await Pubkey2StatesDB.getStates(
+                        pubkeyHash
+                    );
+                    let data = stateIndices.map(async id => {
+                        let state = await l2Storage.state.get(Number(id));
+                        return {
+                            state_id: id,
+                            balance: state.balance.toString(),
+                            token_id: state.tokenID.toString(),
+                            nonce: state.nonce.toString()
+                        };
+                    });
+                    return { states: await Promise.all(data) };
+                } catch (error) {
+                    if (error.name === "NotFoundError") {
+                        return reply
+                            .status(404)
+                            .send({ error: "pubkey not found" });
+                    } else {
+                        return reply.status(501);
```

Comment:
```suggestion
                        return reply.status(500);
```

I think if you just re-throw here fastify will properly set the error to `HTTP 500`.